qemu/tests/balloon_boot_in_pause.py

```diff
@@ -0,0 +1,215 @@
+
+    """
+    Basic functions of memory ballooning test for guest booted
+    in paused status
+    """
+
+    def __init__(self, test, params, env):
+        super(BallooningTest, self).__init__(test, params, env)
+
+        self.vm = env.get_vm(params["main_vm"])
+        self.ori_mem = self.get_vm_mem(self.vm)
+        self.current_mmem = self.get_ballooned_memory()
+        if self.current_mmem != self.ori_mem:
+            self.balloon_memory(self.ori_mem)
+        self.current_mmem = self.ori_mem
+
+    @error_context.context_aware
+    def memory_check(self, step, ballooned_mem):
+        """
+        Check memory change status in monitor and return memory both
+        in guest and monitor
+        :param step: the check point string
+        :type step: string
+        :param ballooned_mem: ballooned memory in current step
+        :type ballooned_mem: int
+        :return: memory size get from monitor and guest
+        :rtype: tuple
+        """
+        error_context.context("Check memory status %s" % step, logging.info)
+        mmem = self.get_ballooned_memory()
+        gmem = self.get_memory_status()
+        if (abs(mmem - self.ori_mem)) != ballooned_mem:
+            self.error_report(step, self.ori_mem - ballooned_mem, mmem)
+            raise exceptions.TestFail("Balloon test failed %s" % step)
+        return (mmem, gmem)
+
+    @error_context.context_aware
+    def balloon_memory(self, new_mem):
+        """
+        Baloon memory for guest in paused status to new_mem and
+        verifies on qemu monitor if the change worked.
+
+        :param new_mem: New desired memory.
+        :type new_mem: int
+        """
+        error_context.context("Change VM memory to %s" % new_mem, logging.info)
+        try:
+            self.vm.balloon(new_mem)
+        except Exception, e:
+            if self.vm.monitor.verify_status('paused'):
+                logging.info("Resume the paused guest")
+                self.vm.monitor.cmd("cont")
+            elif new_mem == self.get_ballooned_memory():
+                pass
+            else:
+                raise exceptions.TestFail("Balloon memory fail with error "
+                                          "message: %s" % e)
+        compare_mem = new_mem
+        balloon_timeout = float(self.params.get("balloon_timeout", 100))
+        status = utils_misc.wait_for((lambda: compare_mem ==
+                                      self.get_ballooned_memory()),
+                                     balloon_timeout)
+        if status is None:
+            raise exceptions.TestFail("Failed to balloon memory to expect"
+                                      " value during %ss" % balloon_timeout)
+
+    def get_memory_boundary(self):
+        """
+        Get the legal memory boundary for the balloon test.
+
+        :return: min and max size of the memory
+        :rtype: tuple
+        """
+        ratio = float(self.params.get("ratio", 0.5))
+        min_size = int(ratio * self.ori_mem)
```

Comment:
@huangyum Updated to use ratio to be 0.8 for guest started with memory less than 1G.